File: programmers/이진탐색/Python/퍼즐 게임 챌린지.py
def puzzle(diffs, times, limit, level) :
    clearTime = 0
    for idx in range(len(diffs)) :
        if diffs[idx] <= level : clearTime += times[idx]
        if diffs[idx] > level : clearTime += (diffs[idx]-level) * (times[idx-1] + times[idx]) + times[idx]
        if clearTime > limit : return False
    return True
    
def solution(diffs, times, limit):#diffs는 100000이 나올거야
    start, mid, end = 1, 0, max(diffs)
    
    while(start <= end) :
        mid = (start + end) // 2
        
        if puzzle(diffs, times, limit, mid) : end = mid - 1
        else : start = mid + 1
    
    if puzzle(diffs, times, limit, mid): return mid
    if puzzle(diffs, times, limit, mid + 1): return mid+1
    return mid-1

File: programmers/이진탐색/Js/퍼즐 게임 챌린지.js
function solution(diffs, times, limit) {
  //문제 제한사항 보고 설정
  //max, min, mid 설정
  let max = 100000;
  let min = 1;
  let mid = 0;
  let answer = 0;

  //이분탐색 진행
  while (min <= max) {
    mid = Math.floor((min + max) / 2);
    let spendTime = 0,
      over = false;

    //숙련도에 따른 소요시간 구하기
    for (let i = 0; i < diffs.length; i++) {
      if (mid - diffs[i] < 0) {
        spendTime += (diffs[i] - mid) * (times[i] + times[i - 1]) + times[i];
      } else {
        spendTime += times[i];
      }

      //spendTime이 limit넘었는지 고려
      if (spendTime > limit) {
        over = true;
        break;
      }
    }

    //타임오바가됐는지에 따라 min, max값 조절
    if (over) {
      min = mid + 1;
    } else {
      answer = mid;
      max = mid - 1;
    }
  }

  return answer;
}
